### PR TITLE
py-pytorch: Update to version 2.3.0; Add Python 311 312

### PR DESCRIPTION
**Pull request by essandess**
Depends upon #24097

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS x.y
Xcode x.y / Command Line Tools x.y.z

###### Verification 
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [ ] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

**Comment by essandess** [comment]
@reneeotten Note that PyTorch also has the Python dependency `cmake`:

https://github.com/pytorch/pytorch/blob/a28bfb5ed545f00e6386519f3f92152001a19ddf/pyproject.toml#L9

I would just defer to upstream usage and consider merging and using #24096, rather than editing this out every time it's used by a major project like PyTorch.

**Comment by reneeotten** [comment]
> @reneeotten Note that PyTorch also has the Python dependency `cmake`:
> 
> https://github.com/pytorch/pytorch/blob/a28bfb5ed545f00e6386519f3f92152001a19ddf/pyproject.toml#L9
> 
> I would just defer to upstream usage and consider merging and using #24096, rather than editing this out every time it's used by a major project like PyTorch.

well, if you want to add it to a package manager you sometime have to do a bit of additional work. In this case once remove the `cmake` dependency listed in `pyproject.toml` or figure out why the build system decided that `cmake` is not installed whereas it clearly is there. So no, following what upstream suggest is not always the best option - you know how to resolve this so please just do it if you want this to be merged.

File: python/py-pytorch/files/Werror.patch [hunk 1]
```diff
@@ -1,211 +0,0 @@
---- third_party/fbgemm/CMakeLists.txt.orig	2023-03-31 09:12:01
-+++ third_party/fbgemm/CMakeLists.txt	2023-03-31 09:14:00
-@@ -144,7 +144,6 @@
- else(MSVC)
-   string(APPEND CMAKE_CXX_FLAGS " -Wall")
-   string(APPEND CMAKE_CXX_FLAGS " -Wextra")
--  string(APPEND CMAKE_CXX_FLAGS " -Werror")
```
Comment (reneeotten) [inline python/py-pytorch/files/Werror.patch:7]: this line is still present in `third_party/fbgemm/CMakeLists.txt` (I haven't checked others) - why do you think that patch is no longer needed? We do not want to compile stuff with `-Werror`.

Comment (essandess) [reply python/py-pytorch/files/Werror.patch]: That patch file no longer works with the latest version. I'll have to generate a new one.

**Comment by essandess** [comment]
> figure out why the build system decided that `cmake` is not installed whereas it clearly is there. 

FYSA: `cmake` in this context refers to the Python module of #24096, which isn't there, not the binary `cmake`, which is.

**Comment by reneeotten** [comment]
> > figure out why the build system decided that `cmake` is not installed whereas it clearly is there.
> 
> FYSA: `cmake` in this context refers to the Python module of #24096, which isn't there, not the binary `cmake`, which is.

and that just installs the usual CMake according to the README: `The CMake python wheels provide CMake 3.29.3.`

**Comment by essandess** [comment]
@reneeotten, @mascguy On macOS version prior to 14, there's this build error:

```sh
CMAKE_LIBRARY_PATH='/opt/local/lib:/opt/local/lib/libomp'
…
LIBRARY_PATH='/opt/local/lib'
…
/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-pytorch/py312-pytorch/work/pytorch-2.3.0/build/CMakeFiles/FindOpenMP/OpenMPTryFlag.c:2:10:
fatal error: 'omp.h' file not found

#include <omp.h>

         ^~~~~~~

1 error generated.

ninja: build stopped: subcommand failed.
```

The Portfile explicitly specifies that `LIBRARY_PATH` includes `${prefix}/lib/libomp`, but this is being overwritten for some reason.

Does anyone know why this error occurs on macOS 11–13, but the build works on macOS 14?

```tclsh
    build.env-append \                                                          
                    CMAKE_LIBRARY_PATH=${prefix}/lib:${prefix}/lib/libomp \     
                    LIBRARY_PATH=${prefix}/lib:${prefix}/lib/libomp \           
```

**Comment by essandess** [comment]
@reneeotten I have successful builds on all macOS versions now. The CI errors are simply timeouts, and the logs show successful builds for many Python versions until the runner times out.

Would you please take a look?

**Comment by essandess** [comment]
@reneeotten This PR is ready to merge. I worked hard to ensure that it builds across all Python and CI versions. The remaining errors are simply CI timeouts because PyTorch takes so long to build.

**Pull request merged**